protocol/protocol_signature.re

```diff
@@ -1,22 +1,45 @@
+open Helpers;
 open Crypto;
 
-[@deriving (ord, yojson)]
+[@deriving ord]
 type t = {
   // TODO: what is the name of a signature?
   signature: Signature.t,
   public_key: Wallet.t,
+  address: Address.t,
 };
 let public_key = t => t.public_key;
+let address = t => t.address;
+
+let (to_yojson, of_yojson) = {
+  module Serialized_data = {
+    [@deriving yojson]
+    type t = {
+      signature: Signature.t,
+      public_key: Wallet.t,
+    };
+  };
+
+  let to_yojson = t =>
+    Serialized_data.to_yojson({
+      signature: t.signature,
+      public_key: t.public_key,
+    });
+  let of_yojson = json => {
+    let.ok {signature, public_key} = Serialized_data.of_yojson(json);
+    let address = Address.of_wallet(public_key);
+    Ok({signature, public_key, address});
+  };
+  (to_yojson, of_yojson);
+};
```

Comment:
This was mostly done to avoid sending redundant data over the network(address and key), but also to ensure that the public_key matches the address.